homeassistant/components/compensation/__init__.py

```diff
@@ -106,6 +105,14 @@ async def async_setup(hass, config):
             }
             data[CONF_POLYNOMIAL] = np.poly1d(coefficients)
 
+            unique_id = data.get(CONF_UNIQUE_ID)
+            if unique_id is None:
+                identifier = data[CONF_SOURCE]
+                attribute = data.get(CONF_ATTRIBUTE)
+                data[CONF_UNIQUE_ID] = (
+                    identifier if attribute is None else f"{identifier}.{attribute}"
```

Comment:
We shouldn't try to be smart and calculate the unique id as fallback. Just keep it None if the user doesn't set it.